framework/sandstone.cpp

```diff
@@ -2685,6 +2709,118 @@ static void warn_deprecated_opt(const char *opt)
             program_invocation_name, opt);
 }
 
+
+static bool system_is_idle(float idle_threshold)
+{
+    #ifdef __linux__
+
+    FILE *loadavg;
+    float load_1, load_5, load_15;
+    int ret;
+
+    // Look at loadavg for hints whether the system is reasonably idle or not
+    // any error and we assume it's busy/under load for simplicity
+
+    loadavg = fopen("/proc/loadavg", "r");
+    if (!loadavg)
+        return false;
+
+    ret = fscanf(loadavg, "%f %f %f", &load_1, &load_5, &load_15);
+    fclose(loadavg);
+
+    if (ret != 3)
+       return false;
+
+    if (load_5 <= idle_threshold)
+       return true;
+    #else //__linux__
+    return true;//TO DO implement idle trigger for Windows
+    #endif
+
+    return false;
+}
+
+static void background_scan_init(void)
+{
+    MonotonicTimePoint now = MonotonicTimePoint::clock::now();
+
+    // init timestamps to 1day old - this quickstarts testing on first run
+    if ((sApp->background_scan.init_timestamp == true) && (sApp->service_background_scan == true)) {
+        sApp->background_scan.timestamp.fill(now - sApp->background_scan.time_to_run_next_batch_of_tests);
+    }
+}
+
+static void background_scan_update_load_threshold(void)
+{
+    MonotonicTimePoint now = MonotonicTimePoint::clock::now();
+
+    hours time_from_last_test = 
+        duration_cast<hours>(now - sApp->background_scan.timestamp[sApp->background_scan.timestamp_newest]);
+    // scale our idle threshold value from 0.2 base, to 0.8 after 12h
+    // every hour adds 0.05 to the threshold value
+    sApp->background_scan.load_idle_threshold = 
+        sApp->background_scan.load_idle_threshold_init +
+        (time_from_last_test.count() * sApp->background_scan.load_idle_threshold_inc_val);
+
+    // prevent the idle threshold form rising above 0.8
+    if(sApp->background_scan.load_idle_threshold > sApp->background_scan.load_idle_threshold_max)
+        sApp->background_scan.load_idle_threshold = sApp->background_scan.load_idle_threshold_max;
+}
+
+static void background_scan_wait(void) 
+{
+    MonotonicTimePoint now = MonotonicTimePoint::clock::now();
+    // Don't run tests unless load is low or it's time to run a test anyway
+    while(1) {
+        // wait ~5 mins no matter what
+        sApp->delay_between_tests = sApp->background_scan.minimum_delay_between_tests;
+
+        double wait_deviation_percent = 10.0;
```

Comment:
Note: 10% of an hour is 6 minutes. We may want to lower this deviation, once we get experience.